src/user_functions/WindEnergyPowerLawAuxFunction.C

```diff
@@ -0,0 +1,83 @@
+/*------------------------------------------------------------------------*/
+/*  Copyright 2019 National Renewable Energy Laboratory.                  */
+/*  This software is released under the license detailed                  */
+/*  in the file, LICENSE, which is located in the top-level Nalu          */
+/*  directory structure                                                   */
+/*------------------------------------------------------------------------*/
+
+#include "user_functions/WindEnergyPowerLawAuxFunction.h"
+#include <algorithm>
+
+// basic c++
+#include <cmath>
+#include <vector>
+#include <stdexcept>
+
+namespace sierra{
+namespace nalu{
+
+WindEnergyPowerLawAuxFunction::WindEnergyPowerLawAuxFunction(
+  const unsigned beginPos,
+  const unsigned endPos,
+  const std::vector<double> &params) :
+  AuxFunction(beginPos, endPos)
+{
+  // check size and populate
+  if ( params.size() != 9 )
+    throw std::runtime_error("Realm::setup_initial_conditions: wind_energy_power_law requires 9 params: ");
+  coord_dir_  = int(params[0]);
+  y_offset_ = params[1];
+  y_ref_ = params[2];
+  shear_exp_ = params[3];
+  u_ref_[0] = params[4];
+  u_ref_[1] = params[5];
+  u_ref_[2] = params[6];
+  u_mag_ = std::sqrt(u_ref_[0] * u_ref_[0] + u_ref_[1] * u_ref_[1] + u_ref_[2] * u_ref_[2]);
+  u_min_ = params[7];
+  u_max_ = params[8];
+}
+
+void
+WindEnergyPowerLawAuxFunction::do_evaluate(
+  const double *coords,
+  const double /*time*/,
+  const unsigned /*spatialDimension*/,
+  const unsigned numPoints,
+  double * fieldPtr,
+  const unsigned fieldSize,
+  const unsigned /*beginPos*/,
+  const unsigned /*endPos*/) const
+{
```

Comment:
The floating point ops in this function could be optimized? It appears to be doing the same calculations again and again. `uref_` vector could be normalized in constructor and used here?